storage/src/checker/range_set.rs

```diff
@@ -0,0 +1,169 @@
+// Copyright (C) 2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE.md for licensing terms.
+
+use btree_range_map::{AnyRange, AsRange, RangeSet};
+use std::num::NonZeroU64;
+use std::ops::Bound;
+
+use crate::nodestore::NodeStoreHeader;
+use crate::{CheckerError, LinearAddress};
+
+pub(super) struct LinearAddressRangeSet {
+    range_set: RangeSet<u64>,
+    size: u64,
+}
+
+impl LinearAddressRangeSet {
+    pub(super) fn new(file_size: u64) -> Result<Self, CheckerError> {
+        if file_size < NodeStoreHeader::SIZE {
+            return Err(CheckerError::InvalidFileSize(file_size));
+        }
+
+        Ok(Self {
+            range_set: RangeSet::new(),
+            size: file_size,
+        })
+    }
+
+    pub(super) fn insert_area(
+        &mut self,
+        addr: LinearAddress,
+        size: u64,
+    ) -> Result<(), CheckerError> {
+        let start = addr.get();
+        // end can be larger than the file size: only the node in the stored area is written to disk
+        if start < NodeStoreHeader::SIZE || start > self.size {
+            return Err(CheckerError::AreaOutOfBounds { start: addr, size });
+        }
+        if self.range_set.intersects(start..start + size) {
+            return Err(CheckerError::AreaIntersects { start: addr, size });
+        }
+        self.range_set.insert(start..start + size);
+        Ok(())
+    }
+}
+
+impl IntoIterator for LinearAddressRangeSet {
+    type Item = (LinearAddress, u64);
+    type IntoIter = std::iter::Map<
+        <RangeSet<u64> as IntoIterator>::IntoIter,
+        fn(AnyRange<u64>) -> (LinearAddress, u64),
+    >;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.range_set.into_iter().map(|range| {
+            let Bound::Included(start) = range.start() else {
+                panic!("start of range is not included, this should not happen");
+            };
+            let Bound::Excluded(end) = range.end() else {
+                panic!("end of range is not excluded, this should not happen");
+            };
```

Comment:
I'm confused why we need these checks and panics. Can't we get an iterator that returns a `RangeFromExcludedToIncluded`? https://docs.rs/btree-range-map/0.7.2/btree_range_map/struct.RangeFromExcludedToIncluded.html seems to have one, not sure why these aren't being returned from `into_iter`.

Reply:
Whenever I see panics, particularly when they are checking some value, my first inclination is to see if the type system can be leveraged to avoid the problem in the first place. Maybe it's not possible in this case.